### ISSUE TITLE
[BUG] geom_col() vs geom_segment()

### ISSUE THREAD

**Issue by rsh52** [CHOP-CGTInformatics/ggswim#32]
# Background

Currently, `ggswim()` wraps `geom_col()` and `add_marker()` wraps `geom_point()`/`geom_label()`. This allows for easier legend separation of swim lanes and markers since `geom_col()` can make use of the `fill` aesthetic while `add_marker()` can use the `color` aesthetic. These different layers have allowed for legends like in the README:

![image](https://github.com/CHOP-CGTInformatics/ggswim/assets/38384694/f13415d2-d834-43a0-b572-13f87696ab65)

# The Issue

`geom_col()`/`geom_bar()` only accept `x` and `y` aesthetics and draw their bars relative to 0, making bar ordering relational to `fill` challenging. Additionally, `position` arg specification either forces:

- `"identity"`, leading to potential for data overlap and hiding.
- `"stack"`, forcing users to normalize their data round 0. `geom_col()` stacking doesn't order things appropriately this way either.

`geom_segment()` addresses these issues by allowing for row-wise `xstart`/`xend` delineation. Segments can be ordered easily. However, it is a wrapper of `geom_line()`, which only uses the `color` aesthetic and not `fill`. Legend separation here is much harder.

## Examples

Here is a sample dataset to explain the problem. 

- `stacked_time` represents lead/lag time between `time_from_infusion` values. Ideally, these would make up the stacked blocks for `geom_col()`. In the example below, the fill values are not aligned correctly.
- `xstart` and `xend` represent time point values for the segments passed to `geom_segment()`. In the example below, the fill values are aligned correctly but the legend uses a color layer.

```r
df <- tibble::tribble(
  ~"pt_id", ~"disease_assessment", ~"time_from_infusion", ~"stacked_time", ~"xstart", ~"xend",
  "1", "RD", -5, -4, -5, -1,
  "1", "RD", -1, -1, -1, 0,
  "1", "CR", 10, 10, 0, 10,
  "1", "RD", 15, 5, 10, 15,
  "2", "CR", -10, -8, -10, -2,
  "2", "RD", -2, -2, -2, 0,
  "2", "CR", 5, 5, 0, 5,
  "2", "CRi", 8, 3, 5, 8
)
```

```r
df |> 
  ggplot(aes(x = stacked_time, y = pt_id, fill = disease_assessment)) +
  geom_col(position = "stack", width = .25)
```
![image](https://github.com/CHOP-CGTInformatics/ggswim/assets/38384694/5161f4cf-d925-4519-b249-9251cd5ef474)

```r
df |> 
  ggplot() +
  geom_segment(aes(x = xstart, xend = xend, y = pt_id, color = disease_assessment),
               linewidth = 10)
```
![image](https://github.com/CHOP-CGTInformatics/ggswim/assets/38384694/75b17b16-dc47-4aa7-bc68-e2d0b833f87a)

## Possible Solutions

- Revert to `geom_segment()` and update how `override`s work in `build_ggswim()`
- Create a new `geom_*()` that forces `geom_col()` to behave the way we want

**Comment by tgerke** [comment]
Backing up a bit, I'm a bit confused as to what data structure the user is expected to provide `ggswim()`. Are you requiring a calculation like the `stacked_time` variable, or will `xstart` and `xend` always be present? If it's the latter (that's my vote, since you could set a default `xstart=0` when it's not provided for a basic swimlane), sticking with `geom_segment()` would seem to make the most sense.

**Comment by rsh52** [comment]
> Backing up a bit, I'm a bit confused as to what data structure the user is expected to provide `ggswim()`. Are you requiring a calculation like the `stacked_time` variable, or will `xstart` and `xend` always be present? If it's the latter (that's my vote, since you could set a default `xstart=0` when it's not provided for a basic swimlane), sticking with `geom_segment()` would seem to make the most sense.

@tgerke `df` was just my attempt to show how both methods can be used, and you're right the user will likely have to do some calculation to get a stacked structure like the `stacked_time` column compatible with `geom_col()`. `geom_segment()` is more promising for the exact reason you mention, just becomes challenging when creating and handling the legend. For example:

```r
df2 <- tibble::tribble(
  ~"pt_id", ~"xval", ~"marker",
  "1", -2, "A",
  "1", 6, "A",
  "2", 3, "A",
  "2", 6, "B",
  "2", 8, "B"
)

df |> 
  ggplot() +
  geom_segment(aes(x = xstart, xend = xend, y = pt_id, color = disease_assessment),
               linewidth = 10) +
  geom_point(data = df2, aes(x = xval, y = pt_id, color = marker), size = 5)
```

![image](https://github.com/CHOP-CGTInformatics/ggswim/assets/38384694/c05d6956-77d9-498a-85c2-cf65705c7625)

Now the markers and lanes are all combined, and later we'd like to introduce labels.

I noticed in `ggswimlane` you also used `geom_bar()`. Did you run into similar issues?

**Comment by tgerke** [comment]
> I noticed in ggswimlane you also used geom_bar(). Did you run into similar issues?

lol no because I didn't do any fancy changes in colors within lines like you're doing here 🤠 In hindsight, `geom_segment()` would have probably been a better choice in `ggswimlane` as well.

**Comment by rsh52** [comment]
Recently found the [`ggnewscale` package](https://eliocamp.github.io/ggnewscale/index.html) which comes up in a small section on "[Splitting Legends](https://ggplot2-book.org/scales-guides#splitting-legends)" in the ggplot2 book. This looks like it will do exactly what we want with splitting the `color` aesthetic layers between `geom_segment()` and `geom_point()` which had originally been the reason for using `fill`s/`geom_col()` instead.

**Comment by rsh52** [comment]
Considering this closed with the merge of #36. `ggswim()` has been migrated over to use `geom_segment()` instead of `geom_col()` and any new issues related to the change have been documented.
